fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,943 @@
+/// This codec takes a large number space (u128) and reduces it to a compact number space.
+///
+/// It will find spaces in the number range. For example:
+///
+/// 100, 101, 102, 103, 104, 50000, 50001
+/// could be mapped to
+/// 100..104 -> 0..4
+/// 50000..50001 -> 5..6
+///
+/// Compact space 0..=6 requires much less bits than 100..=50001
+///
+/// The codec is created to compress ip addresses, but may be employed in other use cases.
+use std::{
+    cmp::Ordering,
+    collections::BinaryHeap,
+    convert::{TryFrom, TryInto},
+    io::{self, Write},
+    net::{IpAddr, Ipv6Addr},
+    ops::RangeInclusive,
+};
+
+use common::{BinarySerializable, CountingWriter, VIntU128};
+use ownedbytes::OwnedBytes;
+use tantivy_bitpacker::{self, BitPacker, BitUnpacker};
+
+use crate::column::{ColumnV2, ColumnV2Ext};
+
+pub fn ip_to_u128(ip_addr: IpAddr) -> u128 {
+    let ip_addr_v6: Ipv6Addr = match ip_addr {
+        IpAddr::V4(v4) => v4.to_ipv6_mapped(),
+        IpAddr::V6(v6) => v6,
+    };
+    u128::from_be_bytes(ip_addr_v6.octets())
+}
+
+/// The cost per blank is quite hard actually, since blanks are delta encoded, the actual cost of
+/// blanks depends on the number of blanks.
+///
+/// The number is taken by looking at a real dataset. It is optimized for larger datasets.
+const COST_PER_BLANK_IN_BITS: usize = 36;
+
+/// The range of a blank in value space.
+///
+/// A blank is an unoccupied space in the data.
+/// Ordered by size
```

Comment:
Can you document the condition you are enforcing here?

(Ranges have a len greater than 3)

Reply:
done